File: C++/reinaldo.cpp
#include <iostream>
#include <cctype>
#include <map>
#include "soluciones.h"

/*
* En la solucion de java existia un metodo para eliminar
* espacios sobre una cadena ya existente, esto es claramente
* mas eficiente.
*/
static std::string leer_sin_espacios() {
	std::string linea{};
	char ch = '\0';

	do {
		std::cin.read(&ch, 1);
		if (!std::isspace(ch)) {
			linea += ch;
		}
	} while (ch != '\n');

	return linea;
}

/*
* Utilidades de mapas propias.
*/

template <typename T, typename U>
U get_or_default(std::map<T, U>& m, const T& k, U d) {
	if (m.contains(k)) {
		return m.at(k);
	}

	return d;
}

template <typename T, typename U>
bool compare_map(std::map<T, U>& first, std::map<T, U> second)
{
	U first_value, second_value;
	bool comparation = true;

	if (first.size() != second.size()) {
		return false;
	}

	for (auto it = first.cbegin(); comparation && it != first.cend(); it++) {
		first_value = it->second;
		second_value = get_or_default(second, it->first, 0);

		if (first_value != second_value) {
			comparation = false;
		}
	}

	return comparation;
}

static void casoPrueba() {
	int contador = 0;

	std::string primeraLinea = leer_sin_espacios();
	std::string segundaLinea = leer_sin_espacios();

	// La frecuencia de cada letra en la segunda linea.
	std::map<char, int> segundaFrecuencia{};
	std::map<char, int> cadaFrecuencia{};

	// Alimentamos la segunda frecuencia.
	for (auto it = segundaLinea.cbegin(); it != segundaLinea.cend(); it++) {
		segundaFrecuencia[*it] = get_or_default(segundaFrecuencia, *it, 0) + 1;
	}

	for (int i = 0; i <= primeraLinea.size() - segundaLinea.size(); i++) {
		cadaFrecuencia.clear();

		// Contabilizamos la frecuencia del slice actual.
		for (auto it = primeraLinea.cbegin() + i;
			it != primeraLinea.cbegin() + i + segundaLinea.size(); it++) {
			cadaFrecuencia[*it] = get_or_default(cadaFrecuencia, *it, 0) + 1;
		}

		// Comparo las palabras bajo el criterio: Misma frecuencia de todas las mismas
		// letras en cualquier orden
		if (compare_map(segundaFrecuencia, cadaFrecuencia)) {
			contador++;
		}
	}

	std::cout << contador << '\n';
}

void lanzadores::reinaldo() {
	int n;
	std::cin >> n;
	std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');

	while (n-- > 0) {
		casoPrueba();
	}
}

File: Java/src/reinaldo/Reinaldo.java
package reinaldo;

import java.util.*;

public class Reinaldo {
	private static Scanner in;

	/**
	 * Devuelve un char array sin espacios del texto especificado.
	 */
	public char[] eliminarEspacios(String texto) {
		StringBuilder builder = new StringBuilder();

		char ch;
		for (int i = 0; i < texto.length(); i++) {
			ch = texto.charAt(i);
			if (!Character.isWhitespace(ch)) {
				builder.append(ch);
			}
		}

		return builder.toString().toCharArray();
	}

	public void casoPrueba() {
		int contador = 0;

		char primeraLinea[] = eliminarEspacios(in.nextLine());
		char segundaLinea[] = eliminarEspacios(in.nextLine());

		Map<Character, Integer> frecuenciaSegunda = new HashMap<>();
		Map<Character, Integer> cadaFrecuencia = new HashMap<>();

		// Frecuencia de la segunda linea
		for (int i = 0; i < segundaLinea.length; i++) {
			frecuenciaSegunda.put(segundaLinea[i], frecuenciaSegunda.getOrDefault(segundaLinea[i], 0) + 1);
		}

		// Buscamos una subcadena con la misma frecuencia.
		for (int i = 0; i <= primeraLinea.length - segundaLinea.length; i++) {
			cadaFrecuencia.clear();

			for (int j = i; j < segundaLinea.length + i; j++) {
				cadaFrecuencia.put(primeraLinea[j], cadaFrecuencia.getOrDefault(primeraLinea[j], 0) + 1);
			}

			// Comparo las palabras bajo el criterio: Misma frecuencia de todas las mismas
			// letras en cualquier orden
			if (frecuenciaSegunda.entrySet().equals(cadaFrecuencia.entrySet())) {
				contador++;
			}
		}

		System.out.println(contador);
	}

	public static void main(String args[]) {
		in = new Scanner(System.in);

		Reinaldo app = new Reinaldo();
		int n = in.nextInt();
		in.nextLine();

		for (int i = 0; i < n; i++) {
			app.casoPrueba();
		}
	}
}